Istanbul is finally here!  \o/

sam-falvo opened the pull request:
Squashed commit of the following:

commit b07ba1ed4dabf00ad6ca98b7ca31b131d185be8f
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Tue Jun 4 10:48:48 2013 -0700

```
removing debugging hardwire to lib-cov
```

commit aa2456122f05d9d6478ad6be085cee86c71cef9f
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Mon Jun 3 17:00:29 2013 -0700

```
Missing fs module in coverage.js
```

commit 3333dda7dc19fbd6dbec88b193ce76da6bc978c6
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Mon Jun 3 15:49:35 2013 -0700

```
Debugging -- why u no work all of a sudden?!?!
```

commit 9578824df0af42cc3dded93535c757e5faa1209d
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Mon Jun 3 14:21:41 2013 -0700

```
Removing unnecessary code
```

commit 722db9c0bc2431fc04c48b36309b41d8cca247ed
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Mon Jun 3 13:52:39 2013 -0700

```
Finishing touches and code clean-up before final review

* Reverted hard-wired library dependencies on lib-cov.  Correct behavior
  has been observationally confirmed, so it's time to wrap things up and
  put things back in order.

  As a feature-request, rch requested that we integrate Rewire into
  Whiskey to better support this kind of testing, where we want to use
  the examples with code-coverage.

* Fixed many missing "global variable" leaks.

* Fixed style issues, such as " vs ' for strings.

* Removed unused references to fs module and fs.appendFileSync().

* I believe I did the right thing to list Istanbul as a proper npm
  dependency for Whiskey.
```

commit 70c9209407d78c6d021ba9c1aeceaabdcf76caac
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Mon Jun 3 13:26:51 2013 -0700

```
Fixed abnormally low hit-vs-missing bug in coverage reports
```

commit 281b650346f7b6cc38ce2aea3660aeb11bdbaf3f
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Fri May 31 15:57:04 2013 -0700

```
HTML reporter once again has proper source listing.
```

commit 9d607c2773f2b5ce60d9291558591fbf5bb997ac
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Fri May 31 14:40:40 2013 -0700

```
Coverage reports for CLI reporter working again
```

commit 035f1ae84c9fb373d0057119f71494038b268440
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Thu May 30 18:09:16 2013 -0700

```
typo
```

commit 94525421152a6083a6db583ce7047a2fd5d26088
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Thu May 30 18:03:31 2013 -0700

```
Work done to convert Istanbul coverage objects into jsCoverage objects.

The conversion seems to work, at least from a type-safety point of view.
However, both HTML and CLI reporters indicate 0 lines of code covered.
Using the json reporter, however, clearly indicates otherwise.

To run:
NODE_PATH=$(pwd)/lib-cov bin/whiskey --coverage --coverage-reporter cli --coverage-file /tmp/hooey --coverage-dir /tmp -t example/test-bdd.js
```

commit c05b7f184527f8a441b11288702224aed8a59889
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Thu May 30 14:09:20 2013 -0700

```
Code cleanup prepping for todays editing.
```

commit 5cbfed7503e4c28a134cf80db389fb37c462d91c
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Wed May 29 15:09:24 2013 -0700

```
blah
```

commit dbbaaf4f59beef617d33e6c7a721bf1c3562ef23
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Mon Apr 29 15:41:45 2013 -0700

```
Fixed missing istanbul instrumentation command
```

commit 9ec686ac3f5ab660b31c4c0d400e86dfbb3dcdad
Author: Samuel A. Falvo II sam.falvo@rackspace.com
Date:   Wed Apr 24 16:37:20 2013 -0700

```
Attempt to move instrumentation over to Istanbul.
```

Kami commented on lib/coverage.js:
```diff
@@ -25,12 +25,44 @@ var path = require('path');
+  var jscov = {};
+
+  for (var file in coverage) {
+    filename = path.basename(file);
```
Already pointed this out in the previous PR - Using just the filaneme wont work correctly with projects with have nested directories and potentially multiple files with the same name.

Kami replied:
Besides this thing and some other minor things I've changed in my branch it looks good to me.

Kami commented:
Pushed some minor cleanup and fixes to my merge branch which you can cherry-pick on top of your branch here - https://github.com/cloudkick/whiskey/compare/sam-falvo-integrating-istanbul

Kami commented:
Pushed another commit which re-enables coverage test to my branch - b4bbedce3127f8b4357c6b4e6c0e5c18c4a63fb0

Feel free to just merge my branch into yours.

Commit:
Merge remote-tracking branch 'cloudkick/master' into integrating-istanbul

Commit:
Merge remote-tracking branch 'cloudkick/sam-falvo-integrating-istanbul' into integrating-istanbul

Commit:
Now uses absolute pathnames internally.

Commit:
chdir resolution

Commit:
Restore jscoverage behavior re: path-local filenames

Previously, it used either the basename, which wouldn't work for
hierarchical libraries (required for npm-installed modules, obviously),
or it used full-on absolute pathnames (inconvenient for cutting and
pasting into shell).

Commit:
Clean up coverage display for HTML reporter

Pro:
- HTML view of code coverage shows only lines that Istanbul actually
  instruments.

Con:
- This change alters the figures reported.  A 98-line file with many
  comments, for example, may appear as though it only has 30 lines.

robert-chiniquy commented:
Can probably rebase/ff out the 'merge remote-tracking branch' commits before merge, lmk if i can help with this.

Commit:
Depend upon --chdir if provided; process.cwd() otherwise.

sam-falvo commented:
@robert-chiniquy Yes, I'll need some help with that at some point.  I'd like to get a +1 on this commit first though.  Heh, a lotta good the wip branch did to hide the need for tweaking.

Commit:
Style cleanup

Kami commented:
@sam-falvo Please merge master into your branch and resolve conflicts so I can test the branch and merge it.

sam-falvo commented:
@Kami Will do; gotta run for the moment, but will do so tomorrow at the latest.  Thanks again for your help and review.

robert-chiniquy commented:
let's review:

```
 Kami commented 20 hours ago
@sam-falvo Please merge master into your branch and resolve conflicts so I can test the branch and merge it.
 sam-falvo commented 20 hours ago
@Kami Will do; gotta run for the moment, but will do so tomorrow at the latest. Thanks again for your help and review.
 Kami closed the pull request 20 hours ago
```
